include/deal.II/distributed/cell_weights.h

```diff
@@ -0,0 +1,183 @@
+     *
+     * An object of this class needs to exist for every DoFHandler associated
+     * with the Triangulation we work on to achieve satisfying work balancing
+     * results.
+     *
+     * A small code snippet follows explaining how to achieve each cell
+     * being weighted by its current number of degrees of freedom.
+     * @code
+     * parallel::distributed::CellWeights<dim, spacedim>
+     *   cell_weights(hp_dof_handler);
+     * cell_weights.register_ndofs_weighting();
+     * @endcode
+     * The weighting function can be changed anytime. Even more ambitious
+     * approaches are possible by submitting customized functions, e.g.
+     * @code
+     * cell_weights.register_custom_weighting(
+     *   [](const FiniteElement<dim, spacedim> &active_fe,
+     *      const typename hp::DoFHandler<dim, spacedim>::cell_iterator &)
+     *   -> unsigned int {
+     *   return 1000 * std::pow(active_fe.dofs_per_cell, 1.6);
+     * });
+     * @endcode
+     * The returned value has to be an unsigned integer and is thus limited by
+     * some large number. It is interpreted as the additional computational load
+     * of each cell. See Triangulation::Signals::cell_weight for a discussion on
+     * this topic.
+     *
+     * @note Be aware that this class connects the weight function to the
+     * Triangulation during its construction. If the Triangulation
+     * associated with the DoFHandler changes during the lifetime of the
+     * latter, an assertion will be triggered in the weight_callback() function.
+     * It is recommended to create a separate object in this case and to destroy
+     * the previous one.
+     *
+     * @ingroup distributed
+     * @author Marc Fehling, 2018
+     */
+    template <int dim, int spacedim = dim>
+    class CellWeights
+    {
+    public:
+      /**
+       * Constructor.
+       *
+       * @param[in] dof_handler The hp::DoFHandler which will be used to
+       *    determine each cell's finite element.
+       */
+      CellWeights(const dealii::hp::DoFHandler<dim, spacedim> &dof_handler);
+
+      /**
+       * Destructor.
+       */
+      ~CellWeights();
+
+      /**
+       * Choose a constant weight @p factor on each cell.
+       */
+      void
+      register_constant_weighting(const unsigned int factor = 1000);
+
+      /**
+       * Choose a weight for each cell that is proportional to its number of
+       * degrees of freedom with a factor @p factor.
+       */
+      void
+      register_ndofs_weighting(const unsigned int factor = 1000);
+
+      /**
+       * Choose a weight for each cell that is proportional to its number of
+       * degrees of freedom <i>squared</i> with a factor @p factor.
+       */
+      void
+      register_ndofs_squared_weighting(const unsigned int factor = 1000);
+
+      /**
+       * Register a custom weight for each cell by providing a function as a
+       * parameter.
+       */
+      void
+      register_custom_weighting(
```

Comment:
can you document the argument of this function, and the arguments of the argument?